tools/test/audio/std_utils/fr_test_measure.m

```diff
@@ -87,16 +89,16 @@
         test.m(:,j) = m0 - m_offs;
 
         %% Check pass/fail
-	if test.fr_lo > 0 && test.fr_hi > 0 && ~isempty(test.fr_rp_max_db)
-		idx0 = find(test.f < test.fr_hi);
-		idx1 = find(test.f(idx0) > test.fr_lo);
-		range_db = max(test.m(idx1,j))-min(test.m(idx1,j));
-		test.rp(j) = range_db/2;
-		if test.rp(j) > test.fr_rp_max_db
-			fprintf('Failed response ch%d +/- %f dBpp (max +/- %f dB)\n', ...
-				test.ch(j), test.rp(j), test.fr_rp_max_db);
-			test.fail = 1;
-		end
+        if test.fr_lo > 0 && test.fr_hi > 0 && ~isempty(test.fr_rp_max_db)
+            idx0 = find(test.f < test.fr_hi);
+            idx1 = find(test.f(idx0) > test.fr_lo);
+            range_db = max(test.m(idx1,j))-min(test.m(idx1,j));
+            test.rp(j) = range_db/2;
+            if test.rp(j) > test.fr_rp_max_db
+                fprintf('Failed response ch%d +/- %f dBpp (max +/- %f dB)\n', ...
+                    test.ch(j), test.rp(j), test.fr_rp_max_db);
+                test.fail = 1;
```

Comment:
If `test.fail` is set to `1` everywhere, we have detected a violation. By default, it is set to 0.

Reply:
> test.fail = test.fail + 1 

Would that work and make sense? If not or not right now then forget it, just a thought.

Reply:
_Would that work and make sense?_
`Yes,`

We will not count beyond '1'; once a test fails, it fails. There is no conceivable way for 'test.fail' to be incremented from 1 to 2 for a given suite, and the test is not reattempted.

Reply:
Thank you for bringing the counter increment to attention.

If you have seen anything different than the information presented here, I would like to know because it is likely that I am missing something or that I have an incorrect understanding of the situation, and I will fix the issue.  


Code point where you have counter 

```
for i = 1:n
	cfg = char(configs(i));
	[n_fail, n_pass] = sof_test_perf(cfg);
	n_fail_tot = n_fail_tot + n_fail;
	n_pass_tot = n_pass_tot + n_pass;

	fprintf('Completed: %s\n', cfg);
	fprintf('Passed   : %d\n', n_pass);
	fprintf('Failed   : %d\n\n', n_fail);
end
```

`Good case` 

```
Verdict,          Test case,     Ch1,     Ch2,     Unit
   Pass,               Gain,   -0.17,   -0.11,       dB
   Pass, Frequency response,    0.22,    0.20,   +/- dB
   Pass,    -3 dB frequency,   21.25,   21.25,      kHz
   Pass,     THD+N -20 dBFS,  -39.54,  -37.57,       dB
   Pass,      THD+N -1 dBFS,  -58.18,  -56.17,       dB
   Pass,          dr_db_min,   72.13,       dB

Completed: sof_test_perf_config.m
Passed   : 4
Failed   : 0

Success: Audio quality test passed.
```

`Bad Case` The frequency response and total harmonic distortion tests were both unsuccessful. 
 
```
Verdict,          Test case,     Ch1,     Ch2,     Unit
   Pass,               Gain,   -0.18,   -0.11,       dB
   Fail, Frequency response,     NaN,     NaN,   +/- dB
   Fail,    -3 dB frequency,     NaN,     NaN,      kHz
   Fail,     THD+N -20 dBFS,  -38.06,  -35.93,       dB
   Fail,      THD+N -1 dBFS,  -56.54,  -54.45,       dB
   Pass,          dr_db_min,   72.54,       dB

Completed: sof_test_perf_config.m
Passed   : 2
Failed   : 2

Error: Audio quality test failed.
```

`Bad Case` The frequency response test failed
```
Verdict,          Test case,     Ch1,     Ch2,     Unit
   Pass,               Gain,   -0.18,   -0.12,       dB
   Fail, Frequency response,     NaN,     NaN,   +/- dB
   Fail,    -3 dB frequency,     NaN,     NaN,      kHz
   Pass,     THD+N -20 dBFS,  -47.54,  -46.33,       dB
   Pass,      THD+N -1 dBFS,  -66.32,  -64.98,       dB
   Pass,          dr_db_min,   75.31,       dB

Completed: sof_test_perf_config.m
Passed   : 3
Failed   : 1

Error: Audio quality test failed.
```

Reply:
I found no `n_fail` counter in `fr_test_measure.m`. 

Imagine that `test.rp(j) > test.fr_rp_max_db` AND  `m_over_mask > 0` fail BOTH in `fr_test_input.m`. Then `test.fail = 1` will report a single failure.

If `test.fail = 1` can report only one failure and `test.rp(j) > test.fr_rp_max_db` has already failed then why does the code keep running and checks `m_over_mask > 0`? It's confusing and pointless. `fr_test_measure.m`  should either count all failures or it should stop testing after the first failure.

Reply:
`test.rp(j) > test.fr_rp_max_db` is a programmable parameter that is configured to limit ripple ( peak to peak )


and ` m_over_mask` is derived one

```
                over_mask = test.m(:,j)-mask_hi';
                idx = find(isnan(over_mask) == 0);
                [m_over_mask, io] = max(over_mask(idx));
                if m_over_mask > 0
                        fprintf('Failed upper response mask around %.0f Hz\n', ...
                                test.f(io(1)));
                        test.fail = 1;
                end
```
So, in either case, you can have a failure, one violation by configuring improper mask or test did not follow upper mask limit. I don't believe it's rational to separate them.


I ran tests,  logged data from the analysis and created a frequency response plot 

![image](https://user-images.githubusercontent.com/61411947/190659812-55dbbf3d-cb09-4202-8fab-7cbcdc5c5c08.png)

Reply:
>  I don't believe it's rational to separate them.

What does "separate" mean?

This discussion is totally unrelated to the nature of the tests. It's only about why does the test code keep running and testing when the test has already failed. This is not about audio at all, it's only about programming logic.

Reply:
Sorry, but they are; please investigate the source variable 'test.m' and map the significance.

Reply:
> Sorry, but they are; 

Who is "they"?

Reply:
> > Sorry, but they are;
> 
> Who is "they"?

Who is "they"? 
I mean  `test.fail` 
OK, let me rephrase: the test criteria `they  test.fail` for pass and fail are obtained from the source 'test.m.'

Reply:
> On line 65 and again on line 72 there is:
> 
> ```
>                 test.fail = 1;
>                 return
> ```
> 
> Why no return here?

don't want the test to continue since we don't have data, thus it's returned.

Reply:
OK, I think I finally get it: you want ALL the `fprintf('Failed...` even when the first one has already failed, correct?

So you don't need `test.fail` or some other variable to count the failures in this test because you can always look at the test logs and look at the (number of...) `fprintf`.

That's all what I was wondering about (I don't know anything about the test itself)

Reply:
Yes, that is one of the simplest ways to examine the test logs using `fprintf` output.
Thanks

Reply:
BTW https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes